File: stem-game-backend/stem-game-backend/src/models/Level.js
const { DataTypes } = require('sequelize');
const { sequelize } = require('../config/database');

const Level = sequelize.define('Level', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
  },
  topicId: {
    type: DataTypes.UUID,
    allowNull: false,
    references: {
      model: 'topics',
      key: 'id',
    },
    onDelete: 'CASCADE',
    comment: 'Foreign key to the topic this level belongs to',
  },
  levelNumber: {
    type: DataTypes.INTEGER,
    allowNull: false,
    validate: {
      min: 1,
      max: 20,
    },
    comment: 'Level number within the topic (1-20)',
  },
  code: {
    type: DataTypes.STRING(100),
    allowNull: false,
    unique: true,
    comment: 'Unique code for the level (e.g., "level_p1_light_sources_1")',
  },
  name: {
    type: DataTypes.STRING(150),
    allowNull: false,
    comment: 'Display name of the level (e.g., "Tap the lights")',
  },
  description: {
    type: DataTypes.TEXT,
    allowNull: true,
    comment: 'Detailed description of what the level involves',
  },
  challengeType: {
    type: DataTypes.ENUM(
      'tap_objects',
      'sort_items',
      'path_finding',
      'puzzle',
      'memory_game',
      'matching',
      'sequencing',
      'multiple_choice',
      'drag_drop',
      'interactive_scene'
    ),
    allowNull: false,
    comment: 'Type of challenge/gameplay for this level',
  },
  difficultyLevel: {
    type: DataTypes.ENUM('easy', 'medium', 'hard'),
    allowNull: false,
    defaultValue: 'easy',
    comment: 'Difficulty level of this specific level',
  },
  estimatedDurationMinutes: {
    type: DataTypes.INTEGER,
    allowNull: false,
    defaultValue: 3,
    validate: {
      min: 1,
      max: 10,
    },
    comment: 'Estimated time to complete in minutes',
  },
  storyText: {
    type: DataTypes.TEXT,
    allowNull: true,
    comment: 'Story/narrative text for the level introduction',
  },
  lessonContent: {
    type: DataTypes.TEXT,
    allowNull: true,
    comment: 'Micro-lesson content before the challenge',
  },
  challengeConfig: {
    type: DataTypes.JSONB,
    allowNull: true,
    defaultValue: {},
    comment: 'JSON configuration for the level challenge (objects to find, correct answers, etc.)',
  },
  hints: {
    type: DataTypes.JSONB,
    allowNull: true,
    defaultValue: [],
    comment: 'Array of hints available for this level',
  },
  successMessage: {
    type: DataTypes.STRING(255),
    allowNull: true,
    comment: 'Message shown when level is completed successfully',
  },
  maxStars: {
    type: DataTypes.INTEGER,
    allowNull: false,
    defaultValue: 3,
    validate: {
      min: 1,
      max: 3,
    },
    comment: 'Maximum stars achievable (typically 3)',
  },
  xpReward: {
    type: DataTypes.INTEGER,
    allowNull: false,
    defaultValue: 10,
    validate: {
      min: 1,
    },
    comment: 'Base XP reward for completing the level',
  },
  coinsReward: {
    type: DataTypes.INTEGER,
    allowNull: false,
    defaultValue: 5,
    validate: {
      min: 0,
    },
    comment: 'Base coins reward for completing the level',
  },
  isActive: {
    type: DataTypes.BOOLEAN,
    defaultValue: true,
    allowNull: false,
    comment: 'Whether this level is active and available to users',
  },
}, {
  tableName: 'levels',
  timestamps: true,
  indexes: [
    {
      unique: true,
      fields: ['code'],
    },
    {
      fields: ['topicId', 'levelNumber'],
    },
    {
      fields: ['challengeType'],
    },
    {
      fields: ['difficultyLevel'],
    },
  ],
});

module.exports = Level;

File: stem-game-flutter/stem-game-flutter/lib/features/world/models/level.dart
import 'package:equatable/equatable.dart';

class Level extends Equatable {
  final String id;
  final String topicId;
  final int levelNumber;
  final String code;
  final String name;
  final String description;
  final String challengeType;
  final Map<String, dynamic>? challengeConfig;
  final String? storyText;
  final String? lessonContent;
  final List<String> hints;
  final String? successMessage;
  final String difficultyLevel;
  final int? estimatedDurationMinutes;
  final int maxStars;
  final int xpReward;
  final int coinsReward;
  final bool? isUnlocked;
  final LevelProgress? userProgress;

  const Level({
    required this.id,
    required this.topicId,
    required this.levelNumber,
    required this.code,
    required this.name,
    required this.description,
    required this.challengeType,
    this.challengeConfig,
    this.storyText,
    this.lessonContent,
    required this.hints,
    this.successMessage,
    required this.difficultyLevel,
    this.estimatedDurationMinutes,
    required this.maxStars,
    required this.xpReward,
    required this.coinsReward,
    this.isUnlocked,
    this.userProgress,
  });

  factory Level.fromJson(Map<String, dynamic> json) {
    return Level(
      id: json['id'] as String? ?? '',
      topicId: json['topicId'] as String? ?? '',
      levelNumber: json['levelNumber'] as int? ?? 0,
      code: json['code'] as String? ?? '',
      name: json['name'] as String? ?? 'Unknown Level',
      description: json['description'] as String? ?? '',
      challengeType: json['challengeType'] as String? ?? 'tap_objects',
      challengeConfig: json['challengeConfig'] as Map<String, dynamic>?,
      storyText: json['storyText'] as String?,
      lessonContent: json['lessonContent'] as String?,
      hints: (json['hints'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      successMessage: json['successMessage'] as String?,
      difficultyLevel: json['difficultyLevel'] as String? ?? 'beginner',
      estimatedDurationMinutes: json['estimatedDurationMinutes'] as int?,
      maxStars: json['maxStars'] as int? ?? 3,
      xpReward: json['xpReward'] as int? ?? 0,
      coinsReward: json['coinsReward'] as int? ?? 0,
      isUnlocked: json['isUnlocked'] as bool?,
      userProgress: json['userProgress'] != null
          ? LevelProgress.fromJson(json['userProgress'] as Map<String, dynamic>)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'topicId': topicId,
      'levelNumber': levelNumber,
      'code': code,
      'name': name,
      'description': description,
      'challengeType': challengeType,
      if (challengeConfig != null) 'challengeConfig': challengeConfig,
      if (storyText != null) 'storyText': storyText,
      if (lessonContent != null) 'lessonContent': lessonContent,
      'hints': hints,
      if (successMessage != null) 'successMessage': successMessage,
      'difficultyLevel': difficultyLevel,
      if (estimatedDurationMinutes != null)
        'estimatedDurationMinutes': estimatedDurationMinutes,
      'maxStars': maxStars,
      'xpReward': xpReward,
      'coinsReward': coinsReward,
      if (isUnlocked != null) 'isUnlocked': isUnlocked,
      if (userProgress != null) 'userProgress': userProgress!.toJson(),
    };
  }

  @override
  List<Object?> get props => [
        id,
        topicId,
        levelNumber,
        code,
        name,
        description,
        challengeType,
        challengeConfig,
        storyText,
        lessonContent,
        hints,
        successMessage,
        difficultyLevel,
        estimatedDurationMinutes,
        maxStars,
        xpReward,
        coinsReward,
        isUnlocked,
        userProgress,
      ];
}

class LevelProgress extends Equatable {
  final bool completed;
  final int? stars;
  final int? timeSpentSeconds;
  final int? attempts;

  const LevelProgress({
    required this.completed,
    this.stars,
    this.timeSpentSeconds,
    this.attempts,
  });

  factory LevelProgress.fromJson(Map<String, dynamic> json) {
    return LevelProgress(
      completed: json['completed'] as bool? ?? false,
      stars: json['stars'] as int?,
      timeSpentSeconds: json['timeSpentSeconds'] as int?,
      attempts: json['attempts'] as int?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'completed': completed,
      if (stars != null) 'stars': stars,
      if (timeSpentSeconds != null) 'timeSpentSeconds': timeSpentSeconds,
      if (attempts != null) 'attempts': attempts,
    };
  }

  @override
  List<Object?> get props => [completed, stars, timeSpentSeconds, attempts];
}
